common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -415,24 +429,65 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOExc
+    partitions.remove(appAttemptShuffleId);
+    logger.info("Finalized shuffle {} from Application {}_{}.", msg.shuffleId, msg.appId, msg.attemptId);
     return mergeStatuses;
   }
 
   @Override
   public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
     if (logger.isDebugEnabled()) {
       logger.debug("register executor with RemoteBlockPushResolver {} local-dirs {} "
-        + "num sub-dirs {}", appId, Arrays.toString(executorInfo.localDirs),
-          executorInfo.subDirsPerLocalDir);
+        + "num sub-dirs {} shuffleManager {}", appId, Arrays.toString(executorInfo.localDirs),
+          executorInfo.subDirsPerLocalDir, executorInfo.shuffleManager);
+    }
+    String shuffleManagerMeta = executorInfo.shuffleManager;
+    if (shuffleManagerMeta.contains(":")) {
+      String mergeDirInfo = shuffleManagerMeta.substring(shuffleManagerMeta.indexOf(":") + 1);
+      try {
+        ObjectMapper mapper = new ObjectMapper();
+        MergeDirectoryMeta mergeDirectoryMeta = mapper.readValue(mergeDirInfo, MergeDirectoryMeta.class);
+        if (mergeDirectoryMeta.attemptId == -1) {
+          // When attemptId is 0, there is no attemptId stored in the ExecutorShuffleInfo.
+          // Only the first ExecutorRegister message can register the merge dirs
+          appsPathsInfo.computeIfAbsent(appId, id ->
+            new AppAttemptPathsInfo(appId, mergeDirectoryMeta.attemptId, executorInfo.localDirs,
+              mergeDirectoryMeta.mergeDir, executorInfo.subDirsPerLocalDir));
+        } else {
+          // If attemptId is not -1, there is attemptId stored in the ExecutorShuffleInfo.
+          // The first ExecutorRegister message from the same application attempt will
+          // register the merge dirs in Shuffle Service. Any later ExecutorRegister message
+          // from the same application attempt will not override the merge dirs. But it can
+          // be overridden by ExecutorRegister message from newer application attempt.
+          // Former attempt's shuffle partitions information will also be cleaned up.
+          boolean newAttemptRegistered = false;
+          if (appsPathsInfo.containsKey(appId)
+              && mergeDirectoryMeta.attemptId > appsPathsInfo.get(appId).attemptId) {
+            newAttemptRegistered = true;
+          }
+          appsPathsInfo.compute(appId, (id, appAttemptPathsInfo) -> {
+            if (appAttemptPathsInfo == null
+                || (appAttemptPathsInfo != null && mergeDirectoryMeta.attemptId > appAttemptPathsInfo.attemptId)) {
+              appAttemptPathsInfo = new AppAttemptPathsInfo(appId, mergeDirectoryMeta.attemptId,
+                executorInfo.localDirs, mergeDirectoryMeta.mergeDir, executorInfo.subDirsPerLocalDir);
+            }
+            return appAttemptPathsInfo;
+          });
+          // It is safe to clean up the AppShufflePartitionInfo
+          if (newAttemptRegistered) {
```

Comment:
We should not rely on `newAttemptRegistered` here - but check if `compute` actually added an entry.
For example, something like this:

```suggestion
          AtomicBoolean newAttemptRegistered = new AtomicBoolean(false);
          appsPathsInfo.compute(appId, (id, appAttemptPathsInfo) -> {
            if (appAttemptPathsInfo == null
                || (appAttemptPathsInfo != null && mergeDirectoryMeta.attemptId > appAttemptPathsInfo.attemptId)) {
              appAttemptPathsInfo = new AppAttemptPathsInfo(appId, mergeDirectoryMeta.attemptId,
                executorInfo.localDirs, mergeDirectoryMeta.mergeDir, executorInfo.subDirsPerLocalDir);
              newAttemptRegistered.set(true);
            }
            return appAttemptPathsInfo;
          });
          // It is safe to clean up the AppShufflePartitionInfo
          if (newAttemptRegistered.get()) {
```